Public/Src/Cache/ContentStore/Distributed/NuCache/EventStreaming/EventHubContentLocationEventStore.cs

```diff
@@ -232,41 +236,90 @@ private async Task ProcessEventsAsync(OperationContext context, List<EventData>
-                    () => sendToActionBlockAsync(),
-                    traceOperationStarted: false).TraceIfFailure(context);
+                    async () =>
+                    {
+                        // Augment the result with the information whether the operation was completed synchronously (i.e. the queues were never full)
+                        // or not.
+                        var task = sendToActionBlockAsync();
+                        bool completedSynchronously = task.Status == TaskStatus.RanToCompletion;
+
+                        var result = await task;
+                        result.CompletedSynchronously = completedSynchronously;
+                        return result;
+                    },
+                    traceOperationStarted: false,
+                    extraEndMessage: r => extraEndMessageForSendToActionBlock(r),
+                    caller: "SendToActionBlockAsync").TraceIfFailure(context);
             }
             else
             {
-                await ProcessEventsCoreAsync(new ProcessEventsInput(state, messages), EventDataSerializer);
+                await ProcessEventsCoreAsync(new ProcessEventsInput(state, messages, actionBlockIndex: 0, eventProcessingBlock: null), EventDataSerializer);
             }
 
-            async Task<BoolResult> sendToActionBlockAsync()
+            async Task<SendToActionBlockResult> sendToActionBlockAsync()
             {
+                var tracingInfo = new Dictionary<int, (int maxQueueSize, int numberOfItemsQueued)>();
                 // This local function "sends" a message into an action block based on the sender's hash code to process events in parallel from different machines.
                 // (keep in mind, that the data from the same machine should be processed sequentially, because events order matters).
                 // Then, it creates a local counter for each processing operation to track the results for the entire batch.
                 foreach (var messageGroup in messages.GroupBy(GetProcessingIndex))
                 {
-                    var eventProcessingBlock = _eventProcessingBlocks[messageGroup.Key];
-                    var input = new ProcessEventsInput(state, messageGroup);
+                    int actionBlockIndex = messageGroup.Key;
+                    var eventProcessingBlock = _eventProcessingBlocks![actionBlockIndex];
+                    var input = new ProcessEventsInput(state, messageGroup, actionBlockIndex, eventProcessingBlock);
+
+                    // Tracking the number of items added to an action block with 'actionBlockIndex'
+                    // and the max number of items in this action block.
+                    var info = tracingInfo.GetOrAdd(actionBlockIndex, _ => (maxQueueSize: 0, numberOfItemsQueued: 0));
+                    tracingInfo[actionBlockIndex] = (maxQueueSize: Math.Max(info.maxQueueSize, eventProcessingBlock.InputCount), numberOfItemsQueued: info.numberOfItemsQueued + 1);
+

```

Comment:
How about we just detect non-synchronous completion below and log the index and queue size in that case? As it stands, this composite message for all the queues will be fairly difficult to parse from Kusto.